Index merging

tomwhite opened the pull request:
### Description

This PR is to discuss how to introduce index merging into htsjdk. The idea is that BAI/SBI/TBI indexes can be constructed in parallel parts then merged in a final step. The code is already used successfully in [Disq](https://github.com/disq-bio/disq) to write indexes in parallel on the fly, which is more efficient than a final, serial indexing step.

Disq uses Spark for parallel processing, but the code here does not have any Spark dependencies.

The main additions are:
* Merger classes (BAMIndexMerger, SBIIndexMerger, TabixIndexMerger)
* Utility methods for shifting virtual file pointers by an offset
* Visibility changes

The indexes being merged depend implicitly on the indexed files having been written in parts too (in headerless form). Do we need to add something to do this to htsjdk too?

Any thoughts on how this addition should be structured? E.g. if it should be broken into pieces; thoughts on testing etc.

cc @lbergelson 

### Checklist

- [ ] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

yfarjoun commented:
What is the state of this? is it waiting for a htsspecs pr? let's rebase this and move it forward please!

tomwhite commented:
@yfarjoun it's waiting for a review.

tomwhite commented:
I've rebased this, and also added a `CRAIIndexMerger` as well as some bug fixes for the tabix merging path.

I think the main point for discussion is how to structure these changes with regard to testing. There are no direct unit tests as the testing is in Disq (e.g. https://github.com/disq-bio/disq/blob/master/src/test/java/org/disq_bio/disq/BaiMergingTest.java). They would be hard to move here without also exposing code to produce headerless part files that are the files to be indexed. I'd be interested to hear people's thoughts on this.

lbergelson commented:
@tomwhite What would we need to push to htsjdk in order to run tests on this?  Do we need the ability to generate headerless bam snippets?  Could we push that to htsjdk first and then port this with some tests?

tomwhite commented:
> What would we need to push to htsjdk in order to run tests on this? Do we need the ability to generate headerless bam snippets?

Yes, that's what we need. I will write something to do this so we can test the code in this PR.

tomwhite commented:
@lbergelson @cmnbroad I've now added tests for all the merge pathways, see `BAMMergerTest`, `CRAMMergerTest`, `VCFMergerTest`.

I've left the code for constructing partitioned BAM files (i.e. with header, part files, terminator) in the tests themselves. This is because it's not clear what the API for writing partitioned files should look like in general. It's not a `SAMFileWriter`, since that is inherently sequential. Disq uses Hadoop files to write, and we don't want to bring in that dependency. I'm not sure we want to create an API for this in the absence of another driving use case, so perhaps it's fine to leave this code in tests, and what we need is a clear description of the partitioned format for BAM, CRAM and VCF?

tomwhite commented:
Following discussion with @lbergelson I've documented the file format for partitioned BAM, CRAM, and VCF.

This should be ready for review now.

pshapiro4broad commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,289 @@
+    private static final int UNINITIALIZED_WINDOW = -1;
+
+    private int numReferences = -1;
+    private List<AbstractBAMFileIndex> indexes = new ArrayList<>();
```
Field can be final
```suggestion
    private final List<AbstractBAMFileIndex> indexes = new ArrayList<>();
```

lbergelson commented on src/main/java/htsjdk/variant/variantcontext/writer/VariantContextWriterBuilder.java:
```diff
@@ -484,11 +484,6 @@ else if (STREAM_TYPES.contains(this.outType))
                 writer = createBCFWriter(outPath, outStreamFromFile);
                 break;
             case VCF_STREAM:
-                if (options.contains(Options.INDEX_ON_THE_FLY)) {
```
why does this have to be removed?

tomwhite replied:
So a tabix index can be written when writing a VCF to a stream.

lbergelson commented on src/test/java/htsjdk/samtools/BAMMergerTest.java:
```diff
@@ -0,0 +1,284 @@
+    private final static Path BAM_FILE = new File("src/test/resources/htsjdk/samtools/BAMFileIndexTest/index_test.bam").toPath();
+
+    /**
+     * Writes a <i>partitioned BAM</i>.
```
I think a description of the file structure expected, would be useful to add to the IndexMerger abstract class.  If you move those classes into a new package you could put it as package documentation.

tomwhite replied:
Done. I haven't moved everything into one package for reasons explained above.

lbergelson commented on src/main/java/htsjdk/samtools/BinningIndexBuilder.java:
```diff
@@ -171,8 +182,10 @@ public BinningIndexContent generateIndexContent() {
+                if (fillInUninitializedValues) {
+                    index[i] = lastNonZeroOffset; // not necessary, but C (samtools index) does this
+                }
+                // note, if fillInUninitializedValues is false BAMIndexWriterTest.compareTextual and compareBinary will have to change
```
On further inspection, those methods no longer exist.

tomwhite replied:
Done

lbergelson commented on src/test/java/htsjdk/tribble/index/tabix/TbiEqualityChecker.java:
```diff
@@ -0,0 +1,145 @@
+
+    private BlockCompressedInputStream blockStream;
+
+    public static void assertEquals(Path vcfFile, Path tbiFile1, Path tbiFile2, boolean identical)
```
could you add documentation explaining the case the identical parameter is for

tomwhite replied:
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,254 @@
+        final Bin bin = new Bin(referenceSequence, binNumber);
+        for (Chunk newChunk : allChunks) {
+            // logic is from BinningIndexBuilder#processFeature
+            final long chunkStart = newChunk.getChunkStart();
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,254 @@
+ * Merges BAM index files for (headerless) parts of a BAM file into a single
+ * index file. The index files must have been produced using an uninitialized window (TODO).
+ */
+public class BAMIndexMerger {
```
Made class final. We can't give it a private no-args constructor as it is referenced from the tests (and Disq), and there are two required final args.

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,254 @@
+        }
+        final int numReferences = bais.get(0).getNumberOfReferences();
+        for (AbstractBAMFileIndex bai : bais) {
+            if (bai.getNumberOfReferences() != numReferences) {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,254 @@
+        }
+        final SAMSequenceDictionary dict = header.getSequenceDictionary();
+        final List<AbstractBAMFileIndex> bais = new ArrayList<>();
+        for (SeekableStream baiStream : baiStreams) {
```
The index files are all read into memory (in the test) so each call to `AbstractBAMFileIndex` is a a memory-backed index, so there's probably not much call for a parallel stream. It would be easy to change later if it showed a performance improvement.

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,254 @@
+        Arrays.parallelPrefix(offsets, (a, b) -> a + b); // cumulative offsets
+
+        try (BinaryBAMIndexWriter writer =
+                     new BinaryBAMIndexWriter(numReferences, baiOut)) {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,289 @@
+
+/**
+ * Merges BAM index files for (headerless) parts of a BAM file into a single
+ * index file. The index files must have been produced using an uninitialized window (TODO).
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,289 @@
+        }
+    }
+
+    public static AbstractBAMFileIndex openIndex(SeekableStream stream, SAMSequenceDictionary dictionary) {
```
Changed to use from the test, so `CachingBAMFileIndexOptimized` can be made package private.

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,289 @@
+        final List<Chunk> allChunks = new ArrayList<>();
+        for (Bin b : bins) {
+            if (b.getReferenceSequence() != referenceSequence) {
+                throw new IllegalArgumentException("Bins have different reference sequences");
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMerger.java:
```diff
@@ -0,0 +1,289 @@
+                throw new IllegalArgumentException("Bins have different reference sequences");
+            }
+            if (b.getBinNumber() != binNumber) {
+                throw new IllegalArgumentException("Bins have different numbers");
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMetaData.java:
```diff
@@ -88,6 +88,13 @@
         }
     }
 
+    BAMIndexMetaData(final long firstOffset, final long lastOffset, final int alignedRecords, final int unAlignedRecords) {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexMetaData.java:
```diff
@@ -207,6 +214,19 @@ long getLastOffset() {
+     * @return a new metadata object shifted by the given offset
+     * @see BlockCompressedFilePointerUtil#shift(long, long)
+     */
+    BAMIndexMetaData shift(final long offset) {
```
It's used by `BAMIndexMerger` so needs to be package private at least.

tomwhite commented on src/main/java/htsjdk/samtools/BAMStreamWriter.java:
```diff
@@ -0,0 +1,122 @@
+/**
+ * Class for writing SAMRecords in BAM format to an output stream.
+ */
+public class BAMStreamWriter {
```
:-)

tomwhite commented on src/main/java/htsjdk/samtools/BAMStreamWriter.java:
```diff
@@ -0,0 +1,122 @@
+    private SAMRecord previousSamRecord;
+    private Chunk previousSamRecordChunk;
+
+    public BAMStreamWriter(OutputStream outputStream, OutputStream indexStream, OutputStream sbiStream, long sbiGranularity, SAMFileHeader header) {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BinningIndexBuilder.java:
```diff
@@ -44,14 +44,17 @@
      *
      * @param referenceSequence
      * @param sequenceLength 0 implies unknown length.  Known length will reduce memory use.
+     * @param fillInUninitializedValues TODO
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/BAMIndexer.java:
```diff
@@ -76,7 +76,18 @@ public BAMIndexer(final File output, final SAMFileHeader fileHeader) {
+     *
+     * @param output     Index will be written here.  output will be closed when finish() method is called.
+     * @param fileHeader header for the corresponding bam file.
+     * @param fillInUninitializedValues TODO
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/CachingBAMFileIndexOptimized.java:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  protected BAMIndexContent query(final int referenceSequence, final int startPos, final int endPos) {
+    seek(4);
```
This is from `AbstractBAMFileIndex`.

tomwhite commented on src/main/java/htsjdk/samtools/CachingBAMFileIndexOptimized.java:
```diff
@@ -0,0 +1,69 @@
+    super(stream, dictionary);
+  }
+
+  protected BAMIndexContent query(final int referenceSequence, final int startPos, final int endPos) {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/CachingBAMFileIndexOptimized.java:
```diff
@@ -0,0 +1,69 @@
+    return super.query(referenceSequence, startPos, endPos);
+  }
+
+  protected void skipToSequence(final int sequenceIndex) {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/CachingBAMFileIndexOptimized.java:
```diff
@@ -0,0 +1,69 @@
+    return super.query(referenceSequence, startPos, endPos);
+  }
+
+  protected void skipToSequence(final int sequenceIndex) {
```
Yes, it's no longer needed as the optimization is now in the superclass.

tomwhite commented on src/main/java/htsjdk/samtools/IndexMerger.java:
```diff
@@ -0,0 +1,44 @@
+    protected final OutputStream out;
+    protected final List<Long> partLengths;
+
+    public IndexMerger(final OutputStream out, final long headerLength) {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/IndexMerger.java:
```diff
@@ -0,0 +1,44 @@
+        this.partLengths.add(headerLength);
+    }
+
+    public abstract void processIndex(final T index, final long partLength);
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/IndexMerger.java:
```diff
@@ -0,0 +1,44 @@
+
+    public abstract void processIndex(final T index, final long partLength);
+
+    public abstract void finish(final long dataFileLength) throws IOException;
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/CachingBAMFileIndexOptimized.java:
```diff
@@ -0,0 +1,69 @@
+ * A subclass of CachingBAMFileIndex that is optimized for merging, by returning
+ * null BAMIndexContent objects if all bins are empty.
+ */
+public class CachingBAMFileIndexOptimized extends CachingBAMFileIndex {
```
Renamed.

tomwhite replied:
I tried moving the mergers into one package but it would mean that we have to make lots of classes/methods on `Bin`, `BAMIndexMetadata`, `BAMIndexContent` public, which doesn't seem particularly desirable. Also, it breaks the pattern in htsjdk where packages are used to group formats rather than operations, e.g. a `cram` package and a `variant` package - which goes against having a `merger` package.

tomwhite commented on src/main/java/htsjdk/samtools/cram/CRAIIndexMerger.java:
```diff
@@ -0,0 +1,37 @@
+import java.io.OutputStream;
+import java.util.zip.GZIPOutputStream;
+
+public class CRAIIndexMerger extends IndexMerger<CRAIIndex> {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/cram/CRAIIndexMerger.java:
```diff
@@ -0,0 +1,37 @@
+
+public class CRAIIndexMerger extends IndexMerger<CRAIIndex> {
+
+  private GZIPOutputStream compressedOut;
```
Done

tomwhite commented on src/main/java/htsjdk/tribble/index/tabix/AllRefsTabixIndexCreator.java:
```diff
@@ -0,0 +1,200 @@
+// change is needed to support index merging.
+
+/**
+ * IndexCreator for Tabix.
```
Done

tomwhite commented on src/main/java/htsjdk/tribble/index/tabix/StreamBasedTabixIndexCreator.java:
```diff
@@ -0,0 +1,78 @@
+import java.nio.file.Path;
+import java.util.List;
+
+public class StreamBasedTabixIndexCreator extends AllRefsTabixIndexCreator {
```
Done

tomwhite commented on src/main/java/htsjdk/samtools/IndexMerger.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class IndexMerger<T> {
```
Done

tomwhite commented:
Awesome! Thanks @lbergelson
